libp2p/protocols/mix/sphinx.nim

```diff
@@ -0,0 +1,301 @@
+import results, sequtils
+import ./[config, crypto, curve25519, serialization, tag_manager, utils]
+
+# Define possible outcomes of processing a Sphinx packet
+type ProcessingStatus* = enum
+  Exit # Packet processed successfully at exit
+  Intermediate # Packet processed successfully at intermediate node
+  Duplicate # Packet was discarded due to duplicate tag
+  InvalidMAC # Packet was discarded due to MAC verification failure
+
+# Compute alpha, an ephemeral public value. Each mix node uses its private key and 
+# alpha to derive a shared session key for that hop. 
+# This session key is used to decrypt and process one layer of the packet.
+proc computeAlpha(
+    publicKeys: openArray[FieldElement]
+): Result[(seq[byte], seq[seq[byte]]), string] =
+  if publicKeys.len == 0:
+    return err("No public keys provided")
+
+  var
+    s: seq[seq[byte]] = newSeq[seq[byte]](publicKeys.len)
+    alpha_0: seq[byte]
+    alpha: FieldElement
+    secret: FieldElement
+    blinders: seq[FieldElement] = @[]
+
+  let x = generateRandomFieldElement().valueOr:
+    return err("Generate field element error: " & error)
+
+  blinders.add(x)
+
+  for i in 0 ..< publicKeys.len:
+    if publicKeys[i].len != FieldElementSize:
+      return err("Invalid public key " & $i)
+
+    # Compute alpha, shared secret, and blinder
+    if i == 0:
+      alpha = multiplyBasePointWithScalars([blinders[i]]).valueOr:
+        return err("Multiply base point with scalars error: " & error)
+
+      alpha_0 = fieldElementToBytes(alpha)
+    else:
+      alpha = multiplyPointWithScalars(alpha, [blinders[i]])
```

Comment:
Do we not need `valueOr` here?

Reply:
nah! these functions don't generate errors